pkg/operator/ceph/cluster/watcher.go

```diff
@@ -130,6 +145,333 @@ func (c *clientCluster) onK8sNode(ctx context.Context, object runtime.Object) bo
+	watchForNodeLoss, err := k8sutil.GetOperatorSetting(ctx, c.context.Clientset, opcontroller.OperatorSettingConfigMapName, "ROOK_WATCH_FOR_NODE_FAILURE", "true")
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to get configmap value `ROOK_WATCH_FOR_NODE_FAILURE`.")
+	}
+
+	_, err = c.context.ApiExtensionsClient.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, "networkfences.csiaddons.openshift.io", metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("networkfences.csiaddons.openshift.io CRD not found, skip creating networkFence")
+			return nil
+		}
+		return pkgerror.Wrapf(err, "failed to get networkfences.csiaddons.openshift.io CRD, skip creating networkFence")
+	}
+
+	if strings.ToLower(watchForNodeLoss) != "true" {
+		logger.Debugf("not watching for node failures `ROOK_WATCH_FOR_NODE_FAILURE` is set to %q", watchForNodeLoss)
+		return nil
+	}
+
+	nodeHasOutOfServiceTaint := false
+	for _, taint := range node.Spec.Taints {
+		if taint.Key == v1.TaintNodeOutOfService {
+			nodeHasOutOfServiceTaint = true
+			logger.Debugf("Found taint: Key=%v, Value=%v on node %s\n", taint.Key, taint.Value, node.Name)
+			break
+		}
+
+	}
+
+	if nodeHasOutOfServiceTaint {
+		err := c.fenceNode(ctx, node, cluster)
+		if err != nil {
+			return pkgerror.Wrapf(err, "failed to create network fence for node %q.", node.Name)
+		}
+		return nil
+	}
+
+	logger.Infof("node %s does not have taint %s, trying to remove networkFence CR if exists", node.Name, v1.TaintNodeOutOfService)
+	err = c.unfenceAndDeleteNetworkFence(ctx, *node, cluster)
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to delete network fence for node %q.", node.Name)
+	}
+
+	return nil
+}
+
+func (c *clientCluster) fenceNode(ctx context.Context, node *corev1.Node, cluster *cephv1.CephCluster) error {
+	volumesInuse := node.Status.VolumesInUse
+	if len(volumesInuse) == 0 {
+		logger.Debugf("no volumes in use for node %q", node.Name)
+		return nil
+	}
+	logger.Debugf("volumesInuse %s", volumesInuse)
+
+	rbdVolumesInUse, _ := getCephVolumesInUse(cluster, volumesInuse)
+	if len(rbdVolumesInUse) == 0 {
+		logger.Infof("no rbd `volumesInUse` for node %q", node.Name)
+		return nil
+	}
+
+	listPVs, err := c.context.Clientset.CoreV1().PersistentVolumes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to list PV")
+	}
+
+	rbdPVList, _ := listRBDAndCephFSPV(listPVs, cluster, rbdVolumesInUse)
+	if len(rbdPVList) == 0 {
+		logger.Info("No rbd PVs found on the node")
+		return nil
+	}
+
+	clusterInfo := cephclient.AdminClusterInfo(ctx, cluster.Namespace, cluster.Name)
+
+	// We only need to create the network fence for any one of rbd/cephFS pv.
+	err = c.fenceRbdImage(ctx, node, cluster, clusterInfo, rbdPVList[0])
+	if err != nil {
+		return pkgerror.Wrapf(err, "failed to fence rbd volumes")
+	}
+
+	// else {
```

Comment:
Instead of keeping these cephfs comments, I'd suggest we remove them from this PR. You can save the comments in a separate branch if needed to help later when cephfs is implemented later. It's awkward to see all the comments.